scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala

```diff
@@ -0,0 +1,417 @@
+    case _ => false
+  }
+}
+
+object MacroOpaqueContainer {
+  import java.io._
+  implicit val myContainerOrderedSerializer = new OrderedSerialization[MacroOpaqueContainer] {
+    val intOrderedSerialization = _root_.com.twitter.scalding.macros.Macros.orderedBufferSupplier[Int]
+
+    override def hash(s: MacroOpaqueContainer) = intOrderedSerialization.hash(s.myField) ^ Int.MaxValue
+    override def compare(a: MacroOpaqueContainer, b: MacroOpaqueContainer) = intOrderedSerialization.compare(a.myField, b.myField)
+
+    override def read(in: InputStream) = intOrderedSerialization.read(in).map(MacroOpaqueContainer(_))
+
+    override def write(b: OutputStream, s: MacroOpaqueContainer) = intOrderedSerialization.write(b, s.myField)
+
+    override def compareBinary(lhs: InputStream, rhs: InputStream) = intOrderedSerialization.compareBinary(lhs, rhs)
+  }
+
+  implicit def arbitraryMacroOpaqueContainer: Arbitrary[MacroOpaqueContainer] = Arbitrary {
+    for {
+      aInt <- arb[Int]
+    } yield MacroOpaqueContainer(aInt)
+  }
+
+  def apply(d: Int): MacroOpaqueContainer = new MacroOpaqueContainer(d)
+}
+
+class MacroOpaqueContainer(val myField: Int) {
+}
+
+object Container {
+  implicit def arbitraryInnerCaseClass: Arbitrary[InnerCaseClass] = Arbitrary {
+    for {
+      anOption <- arb[Set[Double]]
+    } yield InnerCaseClass(anOption)
+  }
+
+  type SetAlias = Set[Double]
+  case class InnerCaseClass(e: SetAlias)
+}
+class MacroOrderingProperties extends FunSuite with PropertyChecks with ShouldMatchers with LowerPriorityImplicit {
+  type SetAlias = Set[Double]
+
+  import ByteBufferArb._
+  import Container.arbitraryInnerCaseClass
+
+  def serialize[T](t: T)(implicit orderedBuffer: OrderedSerialization[T]): InputStream =
+    serializeSeq(List(t))
+
+  def serializeSeq[T](t: Seq[T])(implicit orderedBuffer: OrderedSerialization[T]): InputStream = {
+    import JavaStreamEnrichments._
+
+    val baos = new ByteArrayOutputStream
+    t.foreach({ e =>
+      orderedBuffer.write(baos, e)
+    })
+    baos.toInputStream
+  }
+
+  def rt[T](t: T)(implicit orderedBuffer: OrderedSerialization[T]) = {
+    val buf = serialize[T](t)
+    orderedBuffer.read(buf).get
+  }
+
+  def rawCompare[T](a: T, b: T)(implicit obuf: OrderedSerialization[T]): Int = {
+    obuf.compareBinary(serialize(a), serialize(b)).unsafeToInt
+  }
+
+  def oBufCompare[T](a: T, b: T)(implicit obuf: OrderedSerialization[T]): Int = {
+    obuf.compare(a, b)
+  }
+
+  def checkManyExplicit[T](i: List[T])(implicit obuf: OrderedSerialization[T]) = {
+    val serializedA = serializeSeq(i)
+    val serializedB = serializeSeq(i)
+    (0 until i.size).foreach { _ =>
+      assert(obuf.compareBinary(serializedA, serializedB).unsafeToInt === 0)
+    }
+  }
+
+  def checkMany[T: Arbitrary](implicit ord: Ordering[T], obuf: OrderedSerialization[T]) = forAll { i: List[T] =>
+    checkManyExplicit(i)
+  }
+
+  def clamp(i: Int): Int =
+    i match {
+      case x if x < 0 => -1
```

Comment:
there are a few more tests we want to add:

1) transitivity. I added it to the laws in OrderedSerialization: a < b && b < c implies a < c. Same for equality.
2) a sequence of comparisons in the same stream. You do that in checkMany, but only for equality.